src/crypto/ecdh.rs

```diff
@@ -91,70 +89,87 @@ impl Decryptor for SecretKey {
         // derive shared secret
         let shared_secret = our_secret.diffie_hellman(&their_public);
 
-        // Perform key derivation
-        let z = kdf(
-            self.hash,
-            shared_secret.as_bytes(),
-            self.alg_sym.key_size(),
-            &param,
-        )?;
-
         // Peform AES Key Unwrap
         let encrypted_key_len: usize = match mpis[1].first() {
             Some(l) => *l as usize,
             None => 0,
         };
 
-        let mut encrypted_session_key_vec = vec![0; encrypted_key_len];
-        encrypted_session_key_vec[(encrypted_key_len - encrypted_session_key.len())..]
-            .copy_from_slice(encrypted_session_key);
-
-        let mut decrypted_key_padded = aes_kw::unwrap(&z, &encrypted_session_key_vec)?;
-        // PKCS5-style unpadding (PKCS5 is PKCS7 with a blocksize of 8).
-        //
-        // RFC 6637 describes the padding:
-        // a) "The result is padded using the method described in [PKCS5] to the 8-byte granularity."
-        // b) "For example, assuming that an AES algorithm is used for the session key, the sender MAY
-        // use 21, 13, and 5 bytes of padding for AES-128, AES-192, and AES-256, respectively, to
-        // provide the same number of octets, 40 total, as an input to the key wrapping method."
-        //
-        // So while the padding ensures that the length of the padded message is a multiple of 8, the
-        // padding may exceed 8 bytes in size.
-        {
-            let len = decrypted_key_padded.len();
-            let block_size = 8;
-            ensure!(len % block_size == 0, "invalid key length {}", len);
-            ensure!(!decrypted_key_padded.is_empty(), "empty key is not valid");
-
-            // The last byte should contain the padding symbol, which is also the padding length
-            let pad = decrypted_key_padded.last().expect("is not empty");
+        ecdh_decrypt_unwrap_unpad(
+            *shared_secret.as_bytes(),
+            encrypted_session_key,
+            encrypted_key_len,
+            &self.oid,
+            self.alg_sym,
+            self.hash,
+            fingerprint,
+        )
+    }
+}
 
-            // Padding length seems to exceed size of the padded message
-            if *pad as usize > len {
-                return Err(Error::UnpadError);
-            }
+pub fn ecdh_decrypt_unwrap_unpad(
```

Comment:
no need for the `ecdh` prefix, the module is already called `ecdh`

Reply:
also the`unwrap` is a bit unfortunate, as in the rust context this means panic on error

Reply:
also if this becomes public it should have a doc comment describing what exactly it does

Reply:
Good points. I've made adjustments.

I'm torn about my change to use `KeyParams`: It makes the function parameters a little more expressive, but adds a pointless `oid.clone()`. What do you think?

Reply:
more expressive sounds good